src/xdpinstaller/scripts/xdp-setup.ps1

```diff
@@ -0,0 +1,226 @@
+        }
+    }
+}
+
+# Helper to wait for an adapter to start.
+function Wait-For-Adapters($IfDesc, $Count=1, $WaitForUp=$true) {
+    Write-Verbose "Waiting for $Count `"$IfDesc`" adapter(s) to start"
+    $StartSuccess = $false
+    for ($i = 0; $i -lt 100; $i++) {
+        $Result = 0
+        $Filter = { $_.InterfaceDescription -like "$IfDesc*" -and (!$WaitForUp -or $_.Status -eq "Up") }
+        try { $Result = ((Get-NetAdapter | where $Filter) | Measure-Object).Count } catch {}
+        if ($Result -eq $Count) {
+            $StartSuccess = $true
+            break;
+        }
+        Start-Sleep -Milliseconds 100
+    }
+    if ($StartSuccess -eq $false) {
+        Get-NetAdapter | Format-Table | Out-String | Write-Verbose
+        Write-Error "Failed to start $Count `"$IfDesc`" adapters(s) [$Result/$Count]"
+    } else {
+        Write-Verbose "Started $Count `"$IfDesc`" adapter(s)"
+    }
+}
+
+# Helper to uninstall a driver from its inf file.
+function Uninstall-Driver($Inf) {
+    # Expected pnputil enum output is:
+    #   Published Name: oem##.inf
+    #   Original Name:  xdp.inf
+    #   ...
+    $DriverList = pnputil.exe /enum-drivers
+    $StagedDriver = ""
+    foreach ($line in $DriverList) {
+        if ($line -match "Published Name") {
+            $StagedDriver = $($line -split ":")[1]
+        }
+
+        if ($line -match "Original Name") {
+            $infName = $($line -split ":")[1]
+            if ($infName -match $Inf) {
+                break
+            }
+
+            $StagedDriver = ""
+        }
+    }
+
+    if ($StagedDriver -eq "") {
+        Write-Verbose "Couldn't find $Inf in driver list."
+        return
+    }
+
+    cmd.exe /c "pnputil.exe /delete-driver $StagedDriver 2>&1" | Write-Verbose
+    if (!$?) {
+        Write-Verbose "pnputil.exe /delete-driver $Inf ($StagedDriver) exit code: $LastExitCode"
+    }
+}
+
+# Installs the xdp driver.
+if ($Install -eq "xdp") {
+    if (!(Test-Path $XdpInf)) {
+        Write-Error "$XdpInf does not exist!"
+    }
+    if (!(Test-Path $XdpSys)) {
+        Write-Error "$XdpSys does not exist!"
+    }
+    if (!(Test-Path $XdpCat)) {
+        Write-Error "$XdpCat does not exist!"
+    }
+
+    # Uncomment for debugging only: list installed xdp files   
+    #Get-ChildItem -Path "$SystemFolder\xdp*.*" -File -Recurse | ForEach-Object { Write-Verbose $_.FullName }
+
+    Write-Verbose "netcfg.exe -v -l $XdpInf -c s -i ms_xdp"
+    netcfg.exe -v -l $XdpInf -c s -i ms_xdp | Write-Verbose
+    if ($LastExitCode) {
+        Write-Error "netcfg.exe exit code: $LastExitCode"
+    }
+
+    Start-Service-With-Retry xdp
+
+    Write-Verbose "xdp.sys install complete!"  
+}
+
+# Uninstalls the xdp driver.
+if ($Uninstall -eq "xdp") {
+
+    Write-Verbose "netcfg.exe -u ms_xdp"
+    cmd.exe /c "netcfg.exe -u ms_xdp 2>&1" | Write-Verbose
+    if (!$?) {
+        Write-Host "netcfg.exe failed: $LastExitCode"
+    }
+
+    Uninstall-Driver "xdp.inf"
+
+    Cleanup-Service xdp
```

Comment:
I'm curious why this is necessary. Uninstalling the INF doesn't delete the service that was created via install w/ netcfg?

Reply:
I don't remember. It might not stop the service, so this might be here to ensure the service stops. We can clean this up after the fact.

Reply:
As per what above, removed `Cleanup-Service xdp`.